runtime/cudaq/platform/fermioniq/FermioniqBaseQPU.h

```diff
@@ -0,0 +1,233 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include "common/BaseRemoteRESTQPU.h"
+
+namespace cudaq {
+
+class FermioniqBaseQPU : public BaseRemoteRESTQPU {
+public:
+  virtual bool isRemote() override { return true; }
+
+  /// @brief Return true if locally emulating a remote QPU
+  virtual bool isEmulated() override { return false; }
+
+  /// @brief Set the noise model, only allow this for
+  /// emulation.
+  void setNoiseModel(const cudaq::noise_model *model) override {
+    throw std::runtime_error(
+          "Noise modeling is not allowed on this backend");
+  }
+
+  /// Reset the execution context
+  void resetExecutionContext() override {
+
+    cudaq::debug("reset execution context");
+
+    // set the pre-computed expectation value.
+    if (executionContext->name == "observe") {
+      auto expectation =
+          executionContext->result.expectation(GlobalRegisterName);
+      cudaq::debug("got expectation: {}", expectation);
+      executionContext->expectationValue =
+          executionContext->result.expectation(GlobalRegisterName);
+    }
+    executionContext = nullptr;
+  }
+
+  void launchKernel(const std::string &kernelName,
+                    const std::vector<void *> &rawArgs) override {
+    cudaq::info("launching remote rest kernel ({})", kernelName);
+
+    // TODO future iterations of this should support non-void return types.
+    if (!executionContext)
+      throw std::runtime_error(
+          "Remote rest execution can only be performed via cudaq::sample(), "
+          "cudaq::observe(), or cudaq::draw().");
+
+    // Get the Quake code, lowered according to config file.
+    auto codes = getQuakeCodes(kernelName, nullptr, rawArgs);
+    completeLaunchKernel(kernelName, std::move(codes));
+  }
+
+  /// @brief Extract the Quake representation for the given kernel name and
+  /// lower it to the code format required for the specific backend. The
+  /// lowering process is controllable via the configuration file in the
+  /// platform directory for the targeted backend.
+  std::vector<cudaq::KernelExecution>
+  getQuakeCodes(const std::string &kernelName, void *kernelArgs,
+                 const std::vector<void *> &rawArgs) {
```

Comment:
This can be simplified to reuse the base class' `lowerQuakeCode`, for example,

```
cudaq::ExecutionContext defaultContext("sample", 1);
auto* originalContext = executionContext;
if (executionContext->name == "observe")
  executionContext = &defaultContext; // use this context to skip observe ansatz generation
  
codes = lowerQuakeCode(....);

// Restore execution context 
executionContext =  originalContext;
// We can also check that codes.size() == 1
// Add the spin json 
if (executionContext->name == "observe") {
    auto spin = executionContext->spin.value();
    auto user_data = nlohmann::json::object();
    .....
      
    codes[0].user_data  = user_data;
}
```

This would help reduce a lot of duplication in setting up the IR pipeline below.

Reply:
I tried this initially. The problem is that [BaseRemoteRESTQPU::lowerQuakeCode](https://github.com/NVIDIA/cuda-quantum/blob/ebaf92040cc49d1ea91acc25a9d26a792b0c695b/runtime/common/BaseRemoteRESTQPU.h#L478) does some particular things when the executionContext name equals "observe". It splits the circuit into multiple circuits and we don't want that for the FermioniqQPU as we have a different way of calculating the observables. So in my implementation, i just add the module like it would be a "sample" call.

So I had to implement my own function. I named it getQuakeCodes in order to not confuse the compiler.

I agree that a more elegant fix would be to re-use as much code as possible from lowerQuakeCodes. One idea might be to add class methods to BaseRemoteRESTQPU that then can be overwritten in the subclasses. But I guess thats not a trivial change.

Reply:
> I tried this initially. The problem is that [BaseRemoteRESTQPU::lowerQuakeCode](https://github.com/NVIDIA/cuda-quantum/blob/ebaf92040cc49d1ea91acc25a9d26a792b0c695b/runtime/common/BaseRemoteRESTQPU.h#L478) does some very specific things when the executionContext->name == "observe". It splits the circuit into multiple circuits and we don't want that for the FermioniqQPU as we have a different way of calculating the observables. So in my implementation, i add the module in the same way as it would be a "sample" call.


In the [above](https://github.com/NVIDIA/cuda-quantum/pull/2225#discussion_r1786886535) code suggestion, this part of the code was meant to address the above issue

```
cudaq::ExecutionContext defaultContext("sample", 1);
auto* originalContext = executionContext;
if (executionContext->name == "observe")
  executionContext = &defaultContext; // use this context to skip observe ansatz generation
```

In particular, we set the `executionContext` to a placeholder context of name "sample" before calling `BaseRemoteRESTQPU::lowerQuakeCode`.

Reply:
ok clear. will try it out.

Reply:
Hey Thien, I did it and it worked!